internal/util-logging/src/main/scala/sbt/internal/util/ConsoleAppender.scala

```diff
@@ -333,10 +333,14 @@ private object SuperShellLogger {
       out.println(s"$DeleteLine$l")
       if (progress.length > 0) {
         val stripped = ascii.matcher(l).replaceAll("")
-        val isDebugLine =
-          l.startsWith("[debug]") || l.startsWith(s"${scala.Console.RESET}[debug]")
-        // As long as the line isn't a debug line, we can assume it was printed to
-        // the console and reduce the top padding.
+        // This is a hack to exclude debug lines from decreasing the padding count.
+        // The side effect is that when debug is on, they will unfortunately not reduce
+        // the padding even though they are printed to stdout. If a log line contains
+        // the string "debug", it also won't decrease the padding. It gets complicated
+        // comparing strings formatted with ansi codes, so we have to tolerate the hack
+        // for now. Without it, the supershell region can be incorrectly shrunken which
+        // causes stale tasks to persist at the bottom.
+        val isDebugLine = l.contains("debug")
```

Comment:
I am thinking about this and https://github.com/sbt/util/pull/216, and something doesn't sit well with me.
One detail that's been missed is that ConsoleLogger is used to produce file logs as well via https://github.com/sbt/util/blob/c0146ed0ffd5d55c740be819031833a198cf6ccf/internal/util-logging/src/main/scala/sbt/internal/util/MainLogging.scala#L90-L97, there should be an explicit opt-out for supershell rendering. It also means that we can't have global state in singleton object.

If we handle that, hopefully we shouldn't have to do debug level filtering since by log4j design stdout ConsoleAppender should just see stuff above its log level.

Reply:
To clarify, I should've done the supershell opt-out in https://github.com/sbt/util/pull/196.

Reply:
Yeah. I agree that this is something of a hack and I had to do it this way because I didn't understand all of the details of logging. It's hard to follow how logging is used across util and sbt main.

The reason I figured a singleton was ok was because there would only ever be console output writer and that in the vast majority of instances, only a single command is ever run. But I wasn't happy about using it.

I think I have a handle on how to do this without global state. Hopefully I can get something up soon.

Reply:
> I think I have a handle on how to do this without global state. Hopefully I can get something up soon.

ok. I'll be happy to tag team on any of these parts, like making supershell opt-in for ConsoleAppender etc.

Reply:
Ok. I think I got this working. I'll have a PR up soon.